components/feature/mediasession/src/main/java/mozilla/components/feature/mediasession/MediaSessionFeature.kt

```diff
@@ -0,0 +1,51 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.mediasession
+
+import android.content.Context
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.map
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.feature.mediasession.middleware.MediaSessionMiddleware
+import mozilla.components.lib.state.ext.flowScoped
+import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
+
+/**
+ * Feature implementation that handles MediaSession state changes
+ *
+ * @param applicationContext the application's [Context].
+ * @param mediaServiceClass the media service class will handle the media playback state
+ * @param store Reference to the browser store where tab state is located.
+ */
+class MediaSessionFeature(
+    applicationContext: Context,
+    mediaServiceClass: Class<*>,
+    val store: BrowserStore
+) {
+    private val mediaSessionMiddleWare = MediaSessionMiddleware(applicationContext, mediaServiceClass)
+    private var scope: CoroutineScope? = null
+
+    /**
+     * Starts the feature.
+     */
+    fun start() {
+        scope = store.flowScoped { flow ->
+            flow.ifChanged { it.tabs + it.customTabs }.map { state ->
+                state.tabs.filter { it.mediaSessionState != null }
+            }.collect { tabSessionStates ->
+                mediaSessionMiddleWare.process(tabSessionStates)
+            }
+        }
+    }
```

Comment:
This part confuses me: The middleware is not actually a (store) middleware. We either need a feature (that observes the store) or a middleware (that can look at incoming actions and state changes). Both may be an option.

I assume that a Middleware may be the most straight forward: Whenever we see actions that change media state (e.g. adding or remove a media session state) then look at the new state and launch the service?

Reply:
I see.  I'll look into if using a middleware similar to the original media implementation makes more sense.  Thanks,

Reply:
After some investigation, feature implementation works better since it can observe session states directly without the need to use a aggregate state.  Using the actions we will have to keep count of number of active media sessions.